File: scripts/python/review_comments.py
#!/usr/bin/env python3
# ruff: noqa: T201, S603, S607
"""Fetch unresolved PR review comments from CodeRabbit and save to .reviews/prComments.md."""

import json
import re
import subprocess
import sys
from pathlib import Path
from typing import Any

GRAPHQL_QUERY = """query($owner: String!, $repo: String!, $pr: Int!) {
  repository(owner: $owner, name: $repo) {
    pullRequest(number: $pr) {
      reviews(first: 50) {
        nodes {
          id
          databaseId
          author { login }
          body
          reactions(first: 10) {
            nodes { user { login } content }
          }
        }
      }
      reviewThreads(first: 100) {
        nodes {
          id
          isResolved
          comments(first: 1) {
            nodes { author { login } path line body }
          }
        }
      }
    }
  }
}""".replace("\n", " ")


def strip_html_comments(body: str) -> str:
    return re.sub(r"<!--[\s\S]*?-->", "", body).strip()


def run_graphql(query: str, variables: dict[str, str | int]) -> dict[str, Any]:
    args = ["gh", "api", "graphql"]
    for key, value in variables.items():
        args.extend(["-F", f"{key}={value}"])
    args.extend(["-f", f"query={query}"])

    result = subprocess.run(args, capture_output=True, text=True, encoding="utf-8", check=False)
    if result.returncode != 0:
        raise RuntimeError(result.stderr or "GraphQL query failed")
    return json.loads(result.stdout)


def get_current_user() -> str:
    result = subprocess.run(
        ["gh", "api", "user", "-q", ".login"],
        capture_output=True,
        text=True,
        encoding="utf-8",
        check=True,
    )
    return result.stdout.strip()


def get_repo_info() -> dict[str, Any]:
    result = subprocess.run(
        ["gh", "repo", "view", "--json", "owner,name"],
        capture_output=True,
        text=True,
        encoding="utf-8",
        check=True,
    )
    return json.loads(result.stdout)


def get_pr_number() -> int:
    result = subprocess.run(
        ["gh", "pr", "view", "--json", "number", "-q", ".number"],
        capture_output=True,
        text=True,
        encoding="utf-8",
        check=True,
    )
    pr_number = result.stdout.strip()
    if not pr_number:
        raise RuntimeError("No PR found for current branch")
    try:
        return int(pr_number)
    except ValueError as exc:
        raise RuntimeError(f"Invalid PR number: {pr_number}") from exc


def has_user_reacted(review: dict[str, Any], username: str) -> bool:
    return any(
        r.get("user", {}).get("login") == username and r.get("content") == "THUMBS_UP"
        for r in review.get("reactions", {}).get("nodes", [])
    )


def main() -> None:
    repo_info = get_repo_info()
    pr_number = get_pr_number()
    current_user = get_current_user()

    result = run_graphql(
        GRAPHQL_QUERY,
        {
            "owner": repo_info["owner"]["login"],
            "repo": repo_info["name"],
            "pr": pr_number,
        },
    )

    pr_data = result["data"]["repository"]["pullRequest"]

    coderabbit_reviews = [
        r for r in pr_data["reviews"]["nodes"] if r.get("author", {}).get("login") == "coderabbitai" and r.get("body")
    ]

    unreacted_reviews = [r for r in coderabbit_reviews if not has_user_reacted(r, current_user)]

    unresolved_threads = [
        t for t in pr_data["reviewThreads"]["nodes"] if not t["isResolved"] and t["comments"]["nodes"]
    ]

    review_bodies = [
        {"id": r["id"], "body": strip_html_comments(r["body"])}
        for r in unreacted_reviews
        if strip_html_comments(r["body"])
    ]

    has_content = len(unresolved_threads) > 0 or len(review_bodies) > 0

    reviews_dir = Path(".reviews")
    reviews_dir.mkdir(exist_ok=True)

    if not has_content:
        content = f"# Review Comments\n\n**PR:** #{pr_number}\n\nNo unresolved comments.\n"
        (reviews_dir / "prComments.md").write_text(content, encoding="utf-8")
        print("No unresolved comments found on this PR")
        return

    resolve_cmd = (
        'gh api graphql -f query=\'mutation { resolveReviewThread(input: {threadId: "THREAD_ID"}) '
        "{ thread { isResolved } } }'"
    )
    react_cmd = (
        'gh api graphql -f query=\'mutation { addReaction(input: {subjectId: "REVIEW_ID", '
        "content: THUMBS_UP}) { reaction { content } } }'"
    )

    output = f"""# Review Comments

**PR:** #{pr_number}
**Inline threads (unresolved):** {len(unresolved_threads)}
**Review comments (unreacted):** {len(review_bodies)}

To resolve an inline thread after addressing it:
```bash
{resolve_cmd}
```

To mark a review as addressed (adds reaction):
```bash
{react_cmd}
```

"""

    if unresolved_threads:
        output += "---\n\n# Inline Comments (Unresolved)\n\n"
        thread_parts = []
        for t in unresolved_threads:
            comment = t["comments"]["nodes"][0]
            location = f"{comment['path']}:{comment.get('line', '?')}"
            author = comment.get("author", {}).get("login", "unknown")
            thread_parts.append(
                f"## {location}\n\n**Thread ID:** `{t['id']}`\n**Author:** {author}\n\n{comment['body']}"
            )
        output += "\n\n---\n\n".join(thread_parts)

    if review_bodies:
        output += f"\n\n---\n\n# Review Comments ({len(review_bodies)})\n\n"
        review_parts = [f"**Review ID:** `{r['id']}`\n\n{r['body']}" for r in review_bodies]
        output += "\n\n---\n\n".join(review_parts)

    (reviews_dir / "prComments.md").write_text(output, encoding="utf-8")
    print("Saved review comments to .reviews/prComments.md:")
    print(f"  - {len(unresolved_threads)} unresolved inline threads")
    print(f"  - {len(review_bodies)} review comments")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:  # noqa: BLE001
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

File: scripts/typescript/review-comments.mjs
#!/usr/bin/env node
import { execSync, spawnSync } from "node:child_process";
import { mkdirSync, writeFileSync } from "node:fs";

const GRAPHQL_QUERY = `query($owner: String!, $repo: String!, $pr: Int!) {
  repository(owner: $owner, name: $repo) {
    pullRequest(number: $pr) {
      reviews(first: 50) {
        nodes {
          id
          databaseId
          author { login }
          body
          reactions(first: 10) {
            nodes { user { login } content }
          }
        }
      }
      reviewThreads(first: 100) {
        nodes {
          id
          isResolved
          comments(first: 1) {
            nodes { author { login } path line body }
          }
        }
      }
    }
  }
}`.replace(/\n/g, " ");

// Strip HTML comments from review body
function stripHtmlComments(body) {
	return body.replace(/<!--[\s\S]*?-->/g, "").trim();
}

function runGraphQL(query, variables) {
	const args = ["api", "graphql"];
	for (const [key, value] of Object.entries(variables)) {
		args.push("-F", `${key}=${value}`);
	}
	args.push("-f", `query=${query}`);

	const result = spawnSync("gh", args, { encoding: "utf-8" });
	if (result.status !== 0) {
		throw new Error(result.stderr || "GraphQL query failed");
	}
	try {
		return JSON.parse(result.stdout);
	} catch {
		throw new Error(`Failed to parse GraphQL response: ${result.stdout.slice(0, 200)}`);
	}
}

function getCurrentUser() {
	const result = execSync("gh api user -q .login", { encoding: "utf-8" });
	return result.trim();
}

function getRepoInfo() {
	return JSON.parse(
		execSync("gh repo view --json owner,name", { encoding: "utf-8" }),
	);
}

function getPrNumber() {
	const result = execSync("gh pr view --json number -q .number", {
		encoding: "utf-8",
	}).trim();
	if (!result) {
		throw new Error("No PR found for current branch");
	}
	return result;
}

// Check if user has reacted to a review
function hasUserReacted(review, username) {
	return review.reactions.nodes.some(
		(r) => r.user?.login === username && r.content === "THUMBS_UP",
	);
}

// Add thumbs up reaction to a review
function addReactionToReview(reviewId) {
	const mutation = `mutation($id: ID!) {
		addReaction(input: {subjectId: $id, content: THUMBS_UP}) {
			reaction { content }
		}
	}`.replace(/\n/g, " ");

	return runGraphQL(mutation, { id: reviewId });
}

function main() {
	const command = process.argv[2];

	const repoInfo = getRepoInfo();
	const prNumber = getPrNumber();
	const currentUser = getCurrentUser();

	// Fetch data
	const result = runGraphQL(GRAPHQL_QUERY, {
		owner: repoInfo.owner.login,
		repo: repoInfo.name,
		pr: prNumber,
	});

	const prData = result.data.repository.pullRequest;

	// Get CodeRabbit reviews that user hasn't reacted to yet
	const coderabbitReviews = prData.reviews.nodes.filter(
		(r) => r.author?.login === "coderabbitai" && r.body,
	);

	const unreactedReviews = coderabbitReviews.filter(
		(r) => !hasUserReacted(r, currentUser),
	);

	// Default: fetch and display comments
	// Get ALL unresolved inline threads (any author)
	const unresolvedThreads = prData.reviewThreads.nodes.filter(
		(t) => !t.isResolved && t.comments.nodes[0],
	);

	// Get full body from unreacted CodeRabbit reviews (stripped of HTML comments)
	const reviewBodies = unreactedReviews
		.map((r) => ({ id: r.id, body: stripHtmlComments(r.body) }))
		.filter((r) => r.body.length > 0);

	const hasContent = unresolvedThreads.length > 0 || reviewBodies.length > 0;

	if (!hasContent) {
		mkdirSync(".reviews", { recursive: true });
		writeFileSync(
			".reviews/prComments.md",
			`# Review Comments\n\n**PR:** #${prNumber}\n\n✅ No unresolved comments.\n`,
		);
		console.log("No unresolved comments found on this PR");
		return;
	}

	// Build output
	let output = `# Review Comments

**PR:** #${prNumber}
**Inline threads (unresolved):** ${unresolvedThreads.length}
**Review comments (unreacted):** ${reviewBodies.length}

To resolve an inline thread after addressing it:
\`\`\`bash
gh api graphql -f query='mutation { resolveReviewThread(input: {threadId: "THREAD_ID"}) { thread { isResolved } } }'
\`\`\`

To mark a review as addressed (adds 👍 reaction):
\`\`\`bash
gh api graphql -f query='mutation { addReaction(input: {subjectId: "REVIEW_ID", content: THUMBS_UP}) { reaction { content } } }'
\`\`\`

`;

	// Inline threads section
	if (unresolvedThreads.length > 0) {
		output += `---\n\n# Inline Comments (Unresolved)\n\n`;
		output += unresolvedThreads
			.map((t) => {
				const comment = t.comments.nodes[0];
				const location = `${comment.path}:${comment.line ?? "?"}`;
				const author = comment.author?.login ?? "unknown";
				return `## ${location}\n\n**Thread ID:** \`${t.id}\`\n**Author:** ${author}\n\n${comment.body}`;
			})
			.join("\n\n---\n\n");
	}

	// Review bodies section (full text from unreacted reviews)
	if (reviewBodies.length > 0) {
		output += `\n\n---\n\n# Review Comments (${reviewBodies.length})\n\n`;
		output += reviewBodies
			.map((r) => `**Review ID:** \`${r.id}\`\n\n${r.body}`)
			.join("\n\n---\n\n");
	}

	// Write to file
	mkdirSync(".reviews", { recursive: true });
	writeFileSync(".reviews/prComments.md", output);
	console.log(`Saved review comments to .reviews/prComments.md:`);
	console.log(`  - ${unresolvedThreads.length} unresolved inline threads`);
	console.log(`  - ${reviewBodies.length} review comments`);
}

try {
	main();
} catch (error) {
	console.error("Error:", error.message);
	process.exit(1);
}
